### PR TITLE
[cli][metro-config] fix dom components on windows

### LINKED ISSUE #32998: [SDK 52] Expo DOM Components not showing up on Expo Go and Development Builds
### Minimal reproducible example

https://github.com/StefSchalks/app-52-2

### What platform(s) does this occur on?

Android, iOS

### Where did you reproduce the issue?

in a development build, in Expo Go

### Summary

I was trying Expo DOM Components. What happens is I get a full height white area without any content in it (see image) while the content expected result is a "Hello World" h1 shown in the second image. The second image is from an Android build with the following specifications:
```
"preview": {
      "android": {
        "buildType": "apk"
      }
    },
```

> The screenshot of the version in Expo Go and development builds:
Screenshot_20241114_205235_Expo_Go

> The screenshot of the version in build android version:
Screenshot_20241117_180843_app-52-2

### Environment

```text
expo-env-info 1.2.1 environment info:
    System:
      OS: Windows 11 10.0.22631
    Binaries:
      Node: 20.18.0 - C:\Program Files\nodejs\node.EXE
      npm: 10.8.2 - C:\Program Files\nodejs\npm.CMD
    IDEs:
      Android Studio: AI-242.23339.11.2421.12550806
    npmPackages:
      expo: ^52.0.0 => 52.0.7
      react: 18.3.1 => 18.3.1
      react-dom: 18.3.1 => 18.3.1
      react-native: 0.76.2 => 0.76.2
      react-native-web: ~0.19.13 => 0.19.13
    Expo Workflow: managed
```

### Expo Doctor Diagnostics

```text
Enabled experimental React Native Directory checks. Unset the EXPO_DOCTOR_ENABLE_DIRECTORY_CHECK environment variable to disable this check.
✔ Check package.json for common issues
✔ Check Expo config for common issues
✔ Check native tooling versions
✔ Check if the project meets version requirements for submission to app stores
✔ Check for common project setup issues
✔ Check dependencies for packages that should not be installed directly
✔ Check for app config fields that may not be synced in a non-CNG project
✔ Check npm/ yarn versions
✔ Check for issues with Metro config
✖ Validate packages against React Native Directory package metadata
✔ Check Expo config (app.json/ app.config.js) schema
✔ Check that native modules do not use incompatible support packages
✔ Check for legacy global CLI installed locally
✔ Check that packages match versions required by installed Expo SDK
✔ Check that native modules use compatible support package versions for installed Expo SDK
```

### PR DESCRIPTION
**Pull request by Kudo**
# Why

fixes windows support for dom components
fixes #32998

# How

- the relative import path is broken from windows backslashes. this pr introduces `toPosixPath()` to convert a path as POSIX style.
- the original `fileURLToFilePath` breaks on windows. it turns `file:///C:\Users` to `/C:\Users` which should be `C:\Users`. we can leverage [`URL.fileURLToPath`](https://nodejs.org/api/url.html#urlfileurltopathurl-options) from nodejs. it supports windows and will decode under the hood.
- the transform-worker failed to match filename from `.match(/expo\/dom\/entry\.js/)` because the filename passing from [metro](https://github.com/facebook/metro/blob/6859249a1d5cac1bb4a636d2a535a7a4d667434c/packages/metro/src/DeltaBundler/Transformer.js#L104) could contain backslashes. this pr doesn't fix [rsc related code](https://github.com/expo/expo/blob/d55c4ee44b50e0ddd946145922ede96f77607c7f/packages/%40expo/metro-config/src/transform-worker/transform-worker.ts#L62) btw.

# Test Plan

- [x] ci passed
- [x] windows + android expo go
- [x] windows + prebuild + android release build
- [x] router-e2e + ios expo go (check without regression)
- [x] router-e2e + prebuild + ios release build (check without regression)

# Checklist

- [x] Documentation is up to date to reflect these changes (eg: https://docs.expo.dev and README.md).
- [x] Conforms with the [Documentation Writing Style Guide](https://github.com/expo/expo/blob/main/guides/Expo%20Documentation%20Writing%20Style%20Guide.md)
- [x] This diff will work correctly for `npx expo prebuild` & EAS Build (eg: updated a module plugin).

### CODE REVIEW

File: packages/@expo/cli/src/start/server/metro/MetroBundlerDevServer.ts [hunk 1]
```diff
@@ -802,7 +802,7 @@ export class MetroBundlerDevServer extends BundlerDevServer {
     };
 
     // https://github.com/facebook/metro/blob/2405f2f6c37a1b641cc379b9c733b1eff0c1c2a1/packages/metro/src/lib/parseOptionsFromUrl.js#L55-L87
-    if (!opts.mainModuleName.startsWith('/')) {
+    if (!opts.mainModuleName.startsWith('/') && !path.isAbsolute(opts.mainModuleName)) {
```
Comment (Kudo) [inline packages/@expo/cli/src/start/server/metro/MetroBundlerDevServer.ts:805]: `opts.mainModuleName` can be `C:\Users\...`. without this change, the path would be `./C:\Users\...`, and metro cannot resolve the path

File: packages/@expo/metro-config/src/utils/filePath.ts [hunk 2]
```diff
@@ -0,0 +1,8 @@
+const REGEXP_REPLACE_SLASHES = /\\/g;
+
+/**
+ * Convert any platform-specific path to a POSIX path.
+ */
+export function toPosixPath(filePath: string): string {
+  return filePath.replace(REGEXP_REPLACE_SLASHES, '/');
+}
```
Comment (Kudo) [inline packages/@expo/metro-config/src/utils/filePath.ts:8]: duplicate the function in both cli and metro-config. it doesn't make sense for metro-config to export this utility function for cli use.

File: packages/@expo/cli/src/start/server/metro/MetroBundlerDevServer.ts (hunk 1 above)
Comment (byCedric) [reply packages/@expo/cli/src/start/server/metro/MetroBundlerDevServer.ts:805]: I'm not sure about this, but shouldn't the `mainModuleName` be a module specifier? As in, not be an absolute path, but the path of either `serverRoot` or `projectRoot` + `mainModuleName`

**Comment by byCedric** [comment]
I can confirm this works on Windows, tested it on Windows 11, Node 20, Android Studio Ladybug

**Review by EvanBacon — APPROVED** [review]

File: packages/@expo/cli/src/start/server/metro/MetroBundlerDevServer.ts (hunk 1 above)
Comment (Kudo) [reply packages/@expo/cli/src/start/server/metro/MetroBundlerDevServer.ts:805]: it's happening from `npx expo export:embed` and the mainModuleName would in some cases passing an absolute path like `C:\Users\...`.

i'm keeping this change and merging the pr. if there's any problems, i can follow up in a separate pr.

**Pull request merged**